compiler/src/dmd/dstruct.d

```diff
@@ -350,8 +350,14 @@ extern (C++) class StructDeclaration : AggregateDeclaration
 
         // Determine if struct is all zeros or not
         zeroInit = true;
+        auto lastOffset = -1;
         foreach (vd; fields)
         {
+            // only consider first sized member of an (anonymous) union
+            if (vd.overlapped && vd.offset == lastOffset && vd.type.size(vd.loc) != 0)
+                continue;
+            lastOffset = vd.offset;
```

Comment:
Nice! - I meant the *previous* field's size though; I guess this fails (wrongly treated as zero-initialized):
```D
struct {
    int[0] dummy; // same offset as anon union, but doesn't overlap; should be ignored anyway for zero-init check
    union {
        float f; // is the first member of the anon union and must be checked
        int i;
    }
}
```

Reply:
[So a simple early `if (vd.type.size(vd.loc) == 0) continue;` at the iteration start should do the job, shortcutting the check for 0-length static arrays and *not* setting `lastOffset`, so that the first union member is always checked.]

Reply:
Oops, thanks. Added that test. It is a lot better, but there are (less common) cases where it wrongly doesn't zero init:
```d
union U3 {
    int x;
    union {
        int y;
        struct {
            float z, w; // z ignored but w has different offset to x
        }
    }
}
```
If this is merged, I'll file that case separately.

Reply:
Oof yeah, no idea how we'd easily fix those cases.

Reply:
Filed: https://issues.dlang.org/show_bug.cgi?id=24797